docs/key-issues/multiple_architectures.md

```diff
@@ -0,0 +1,157 @@
+issued by the binary.
+
+Historically, it could be assumed that an executable or library would be
+compiled for a single CPU archicture. On the rare occasion that an operating
+system was available for mulitple CPU architectures, it became the
+responsibility of the user to find (or compile) a binary that was compiled for
+their host CPU architecture.
+
+However, on occasion, we see an operating system platform where multiple CPU
+architectures are supported:
+
+* In the early days of Windows NT, both x86 and DEC Alpha CPUs were supported
+* Windows 10 supports x86, x86-64, ARMv7 and ARM64; Windows 11 supports x86-64
+  and ARM64.
+* Although Linux started as an x86 project, the Linux kernel is now available a
+  wide range of other CPU architectures, including ARM64, RISC-V, PowerPC, s390
+  and more.
+* Apple transitioned Mac hardware from PowerPC to Intel (x86-64) CPUs, providing
+  a forwards compatibility path for binaries
+* Apple is currently transitioning Mac hardware from Intel (x86-64) to
+  Apple Silicon (ARM64) CPUs, again providing a forwards compatibility
+  path
+* Apple supports ARMv6, ARMv7, ARMv7s, ARM64 and ARM64e on iOS
+* Android currently supports ARMv7, ARM64, x86, and x86-64; it has historically
+  also supported ARMv5 and MIPS
+
+CPU architecture compatibility is a necessary, but not sufficient criterion for
+determining binary compatibility. Even if two binaries are compiled for the same
+CPU architecture, that doesn't guarantee [ABI compatibility](abi.md).
+
+In some respects, CPU architecture compatibility could be considered a superset
+of [GPU compatibility](gpus.md). When dealing with multiple CPU architectures,
+there may be some overal with the solutions that can be used to support GPUs in
+native binaries.
+
+## Platform approaches for dealing with multiple architectures
+
+Three approaches have emerged for handling multiple CPU architectures.
+
+### Multiple binaries
+
+The minimal solution is to distribute multiple binaries. This is the approach
+that was used by Windows NT, and is currently supported by Linux. At time of
+distribution, an installer or other downloadable artefact is provided for each
+supported platform, and it is up to the user to select and download the correct
+artefact.
+
+### Archiving
+
+The approach taken by Android is very similar to the multiple binary approach,
+with some affordances and tooling to simplify distribution.
+
+When building an Android project, each target architecture is compiled
+independently. If a native binary library is required to compile the Android
+application, a version must be provided for each supported CPU architecture. A
+directory layout convention exists for providing a binary for each platform,
+with the same library name. This yields an independent final binary (APK) for
+each CPU architecture. When running locally, a CPU-specific APK will be
+uploaded to the simulator or test device.
```

Comment:
The Android native compilers do indeed build each architecture independently, and produce a separate set of .so files for each one. However, it's not correct that the developer builds a separate APK for each architecture. In fact, an APK, like an AAB, contains a directory structure with a subdirectory for each architecture. And before AABs were introduced, the standard way of releasing an Android app was with a multi-architecture APK.

So with Chaquopy, the APK build process contains two steps where native code is involved:
* The standard Android support for building native libraries. Because Android is based on Java/Kotlin, most apps don't use this step.
* Chaquopy's execution of pip, as detailed below. This works with pre-compiled wheels containing .so files, and never executes any compilers.

These two build steps are independent: one does not happen within the other. And each one has its own "for each ABI" loop.

Reply:
Ah - thanks for the clarification. Not sure where I got mislead about APKs being single binary - I seem to recall a historical workflow with Google Play Store that involved multiple APK uploads, but I might be (a) misremembering, or (b) have been doing some redundant uploads at the time.